index.js

```diff
@@ -2,24 +2,207 @@
 const { getAllRealms, getRealmAdmins, getRealmSettings } = require('./libs/get-realms');
-const { KC_CONFIG, KC_TERMS } = require('./constants');
+const { KC_CONFIG, KC_TERMS, IDP_TERMS, KC_MIGRATION_ROUTES, MODE } = require('./constants');
 
 const main = async () => {
   try {
     // setup kc admin instance:
     const kcAdminClient = new kcAdmin(KC_CONFIG.REQUEST);
     await kcAdminClient.auth(KC_CONFIG.AUTH);
+
+    // ====================================== SSO Service Name Migration ======================================>>>
+    // 0. double check before running!
+
+    const newRoute = KC_MIGRATION_ROUTES.NEW;
+    const oldRoute = KC_MIGRATION_ROUTES.OLD;
+    const questions = [
+      {
+        type: 'text',
+        name: 'confirm',
+        message: '================== Please note that you are running EXECUTION mode, are you sure to proceed? (CtrlC to exit if not)==================',
+      },
+      {
+        type: 'text',
+        name: 'env',
+        message: `We are in ${MODE.ENV} environment now, have you taken a DB backup?`
+      },
+      {
+        type: 'text',
+        name: 'about',
+        message: `We are switching ${oldRoute} to ${newRoute}`,
+      }
+    ];
+
+    if (MODE.EXECUTE_CHANGE) {
+      await prompts(questions);
+    }
+
+    // 1. get all IDP realms, IDPs, and clients:
+
+    const allIdpClients = await IDP_TERMS.reduce(async (acc, r) => {
+      const clientList = await acc;
+      const idpRealmId = r.REALM;
+      const idpRealm = await kcAdminClient.realms.findOne({
+        realm: idpRealmId,
+      });
+      const idpClients = await kcAdminClient.clients.find({
+        realm: idpRealmId,
+      });
+      const idpAlias = idpRealm.identityProviders? idpRealm.identityProviders.map(idp => idp.alias) : [];
+      const idpClientIds = idpClients? idpClients.map(client => client.clientId) : [];
+      // Output for checking:
+      // console.log(`--- \nIDP realm: ${idpRealm.id}`);
+      // console.info(`IDP setup: ${idpAlias}`);
+      // console.log('Clients:');
+      // console.info(idpClientIds);
+
+      // get all clients for APP realms:
+      const appRealmClients = idpClientIds.filter(c => c.includes('https'));
+      clientList.push({
+        realm: r.ALIAS,
+        clients: appRealmClients,
+      });
+      return clientList;
+    }, Promise.resolve([]));
+
+    if (!MODE.EXECUTE_CHANGE) {
+      console.log('------------- all clients from IDP realms:');
+      console.log(allIdpClients);
+    }
+
+    // ------------------------------------------------------------------------------------------------------
+    // 2. get all APP realms (other than IDP realms), and the IDPs in each realm:
+
+    const allRealms = await kcAdminClient.realms.find();
+    const idpRealmNames = IDP_TERMS.map(r => r.REALM);
+    const appRealms = allRealms.reduce((acc, r) => {
+      if (!idpRealmNames.includes(r.realm)) {
```

Comment:
not needing a change but since you are just using appRealms for lookip. You could do a reduce instead of a map earlier to turn the realmnames into an object/lookup table

```js
const idpRealmNames = IDP_TERMS.reduce((obj, idp) => {
    obj[idp.REALM] = idp.REALM;
    return obj;
}, {});

```
then use it like 
` if(!idpRealmNames.hasOwnProperty(r.realm)) { ...}`